### PR TITLE
stdio/format: Fix precision modifier for integers

### PR DESCRIPTION
**Pull request by astalke**
DONE: RTOS-373

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: stdio/format.c [hunk 1]
```diff
@@ -483,7 +479,20 @@ static void format_sprintf_num(void *ctx, feedfunc feed, uint64_t num64, uint32_
...
 	}
 
-	const int digits_cnt = tmp - tmp_buf;
+	if ((flags & FLAG_OCT) != 0 && (num64 != 0 || float_frac_len == 0) && (flags & FLAG_ALTERNATE) != 0) {
+		*tmp++ = '0';
+	}
+
+	for (i = tmp - tmp_buf; i < float_frac_len; ++i) {
+		*tmp++ = '0';
+	}
```
Comment (agkaminski) [inline stdio/format.c:488]: Why not memset?

File: stdio/format.c [hunk 2]
```diff
@@ -483,7 +479,20 @@ static void format_sprintf_num(void *ctx, feedfunc feed, uint64_t num64, uint32_
...
+
+	if ((flags & FLAG_HEX) != 0 && num64 != 0 && (flags & FLAG_ALTERNATE) != 0) {
```
Comment (agkaminski) [inline stdio/format.c:490]: MISRA nitpick
```suggestion
	if (((flags & FLAG_HEX) != 0) && (num64 != 0) && ((flags & FLAG_ALTERNATE) != 0)) {
```

File: stdio/format.c [hunk 3]
```diff
@@ -483,7 +479,20 @@ static void format_sprintf_num(void *ctx, feedfunc feed, uint64_t num64, uint32_
 		}
 	}
 
-	const int digits_cnt = tmp - tmp_buf;
+	if ((flags & FLAG_OCT) != 0 && (num64 != 0 || float_frac_len == 0) && (flags & FLAG_ALTERNATE) != 0) {
```
Comment (agkaminski) [inline stdio/format.c:482]: MISRA loves its parentheses
```suggestion
	if ((flags & FLAG_OCT) != 0 && ((num64 != 0) || (float_frac_len == 0)) && ((flags & FLAG_ALTERNATE) != 0)) {
```

File: stdio/format.c [hunk 4]
```diff
@@ -483,7 +479,21 @@ static void format_sprintf_num(void *ctx, feedfunc feed, uint64_t num64, uint32_
...
+	}
+
+	if (tmp - tmp_buf < float_frac_len) {
+		memset(tmp, '0', float_frac_len - (tmp - tmp_buf));
+		tmp = tmp_buf + float_frac_len;
+	}
+
+	if (((flags & FLAG_HEX) != 0) && (num64 != 0) && ((flags & FLAG_ALTERNATE) != 0)) {
+		memcpy(tmp, prefix, 2);
+		tmp += 2;
```
Comment (kemonats) [inline stdio/format.c:493]:
```suggestion
		tmp[0] = prefix[0];
		tmp[1] = prefix[1];
		tmp += 2;
```
`tmp` and `prefix` are `char` type, it is faster.

**Review by agkaminski — DISMISSED** [review]
@damianloew can you verify that this fixes the issue?

**Review by agkaminski — APPROVED** [review]

**Comment by astalke** [comment]
Referencing connected issues:
 - https://github.com/phoenix-rtos/phoenix-rtos-project/issues/277

**Review by damianloew — APPROVED** [review]
LGTM

**Comment by nalajcie** [comment]
Note:
if You would write in commit message body:
`Fixes phoenix-rtos/phoenix-rtos-project#277`

It would automatically link the PR to issue and close the issue on PR merge.

**Comment by astalke** [comment]
> Note: if You would write in commit message body: `Fixes phoenix-rtos/phoenix-rtos-project#277`
> 
> It would automatically link the PR to issue and close the issue on PR merge.

Technically, that issue is not fixed, because we don't support %n.

**Pull request merged**